src/middlewared/middlewared/plugins/crypto.py

```diff
@@ -101,60 +252,817 @@ def cert_issuer(cert):
+            )
+        except Exception:
+            return None
+        else:
+            return certificate.digest('sha1').decode()
+
+    @private
+    async def san_to_string(self, san_list):
+        san_string = ''
+        ip_validator = IpAddress()
+        for count, san in enumerate(san_list or []):
+            try:
+                ip_validator(san)
+            except ShouldBe:
+                san_string += f'DNS: {san}, '
+            else:
+                san_string += f'IP: {san}, '
+        return san_string[:-2] if san_list else ''
+
+
+    @accepts(
+        Dict(
+            'certificate_cert_info',
+            Int('key_length'),
+            Int('serial', required=False),
+            Int('lifetime', required=True),
+            Str('country', required=True),
+            Str('state', required=True),
+            Str('city', required=True),
+            Str('organization', required=True),
+            Str('common', required=True),
+            Str('email', validators=[Email()], required=True),
+            Str('digest_algorithm', enum=['SHA1', 'SHA224', 'SHA256', 'SHA384', 'SHA512']),
+            List('san', items=[Str('san')]),
+            register=True
+        )
+    )
+    async def create_certificate(self, cert_info):
+
+        def _create_ssl_certificate(cert_info):
+
+            cert = crypto.X509()
+            cert.get_subject().C = cert_info['country']
+            cert.get_subject().ST = cert_info['state']
+            cert.get_subject().L = cert_info['city']
+            cert.get_subject().O = cert_info['organization']
+            cert.get_subject().CN = cert_info['common']
+            # Add subject alternate name in addition to CN
+
+            if cert_info['san']:
+                cert.add_extensions([crypto.X509Extension(
+                    b"subjectAltName", False, cert_info['san'].encode()
+                )])
+                cert.get_subject().subjectAltName = cert_info['san']
+            cert.get_subject().emailAddress = cert_info['email']
+
+            serial = cert_info.get('serial')
+            if serial is not None:
+                cert.set_serial_number(serial)
+
+            cert.gmtime_adj_notBefore(0)
+            cert.gmtime_adj_notAfter(cert_info['lifetime'] * (60 * 60 * 24))
+
+            cert.set_issuer(cert.get_subject())
+            # Setting it to '2' actually results in a v3 cert
+            # openssl's cert x509 versions are zero-indexed!
+            # see: https://www.ietf.org/rfc/rfc3280.txt
+            cert.set_version(2)
+            return cert
+
+        cert_info['san'] = await self.san_to_string(cert_info.pop('san', []))
+
+        return await self.middleware.run_in_io_thread(_create_ssl_certificate, cert_info)
+
+    @accepts(
+        Patch(
+            'certificate_cert_info', 'certificate_signing_request',
+            ('rm', {'name': 'lifetime'})
+        )
+    )
+    async def create_certificate_signing_request(self, cert_info):
```

Comment:
Seems like it would make much more sense to not make this a coroutine

You can call self.san_to_string using `self.middleware.call_sync('certificate.san_to_string')